README.md
[![Morpx-bbs](http://bbs.morpx.com/template/mu/images/logo.png)](http://bbs.morpx.com/forum.php)
[![Git-rep](https://img.shields.io/github/repo-size/mu-opensource/MuVisionSensor3.svg)](https://github.com/mu-opensource/MuVisionSensor3)
[![Git-release](https://img.shields.io/github/downloads/mu-opensource/MuVisionSensor3/total.svg)](https://github.com/mu-opensource/MuVisionSensor3/releases)


MuVisionSensor
==============

[![Pic-vision-sensor](http://mai.morpx.com/images/page201904/banner1.jpg)](http://mai.morpx.com/)

The MU Vision Sensor is a sensor module that supports Arduino, Microbit, and other haredware platform that supports UART or I2C communication protocols.

You can use these libraries to read data or set properties of the MU Vision Sensor.

## Simple example

How quickly can you get up and running with the library?  Here's a simple program:
```cpp
#include <MuVisionSensor.h>
#include <Wire.h>
#define VISION_TYPE VISION_BALL_DETECT
#define MU_ADDRESS    0x60
MuVisionSensor Mu(MU_ADDRESS);
void setup() { 
  Wire.begin();                                           // I2C or Serial must initialize before MU vision sensor
  Mu.begin(&Wire);                                        // MU begin
  Mu.VisionBegin(VISION_TYPE);                            // vision begin
}
void loop() {
  if (Mu.GetValue(VISION_TYPE, kStatus)) {                // get vision status value
    int x = Mu.GetValue(VISION_TYPE, kXValue);            // get vision X axes value
    int y = Mu.GetValue(VISION_TYPE, kYValue);            // get vision Y axes value
    int width = Mu.GetValue(VISION_TYPE, kWidthValue);    // get vision width value
    int height = Mu.GetValue(VISION_TYPE, kHeightValue);  // get vision height value
    int label = Mu.GetValue(VISION_TYPE, kLabel);         // get vision label value
  }
}
```
## More examples

You can get more examples under the folder `examples/`.

If you want to test your vision sensor, you can refer to the examples under the folder `examples/test/`.

Or you want to use vision sensor without the library `MuVisionSensor`, you can refer to the examples under the folder `examples/use_arduino_base_interface_only/`.

## Supported platforms

Right now the library is supported on a variety of arduino compatable platforms. If you want to use the library on other plantforms which supported on C/C++, please follow the steps below:

### Uart

> 1.Edit file `mu_vision_sensor_uart_hw_interface.h` in path `MuVisionSensor/src/` 

> 2.Modify typedef `MuVsUart` in line `21` to the UART Class supported by the platform

> 3.Save and close file `mu_vision_sensor_uart_hw_interface.h`

> 4.Edit file `mu_vision_sensor_uart_hw_interface.cpp` in path `MuVisionSensor/src/`

> 5.Override the functions below:
```cpp
/**
  * @brief  Uart read characters from stream into buffer.
  * @param  temp: uart buffer.
  * @param  length: the number of characters you want to read.
  * @retval returns the number of characters placed in the buffer, terminates if length characters have been read, or timeout.
  */
uint32_t MuVisionSensorUart::UartRead(uint8_t* temp, uint8_t length);

/**
  * @brief  Uart write characters from buffer into stream.
  * @param  temp: uart buffer.
  * @param  length: the number of characters you want to write.
  * @retval returns the number of characters placed in the stream, terminates if length characters have been write, or timeout.
  */
uint32_t MuVisionSensorUart::UartWrite(uint8_t* temp, uint8_t length);
```
> 6.Save and close file `mu_vision_sensor_uart_hw_interface.cpp`

### I2C

> 1.Edit file `mu_vision_sensor_i2c_hw_interface.h` in path `MuVisionSensor/src/` 

> 2.Modify typedef `MuVsI2C` in line `24` to the I2C Class supported by the platform

> 3.Save and close file `mu_vision_sensor_i2c_hw_interface.h`

> 4.Edit file `mu_vision_sensor_i2c_hw_interface.cpp` in path `MuVisionSensor/src/`

> 5.Override the functions below:
```cpp
/**
  * @brief  I2C read byte.
  * @param  reg_address: register address.
  * @param  temp: register value.
  * @retval 0: read success
  *         not 0: error
  */
uint32_t MuVisionSensorI2C::I2CRead(uint8_t reg_address, uint8_t* temp);

/**
  * @brief  I2C write byte.
  * @param  reg_address: register address.
  * @param  value: the value write to register.
  * @retval 0: write success
  *         not 0: error
  */
uint32_t MuVisionSensorI2C::I2CWrite(uint8_t reg_address, uint8_t value);
```
> 6.Save and close file `mu_vision_sensor_i2c_hw_interface.cpp`

## What about other library for MuVisionSensor?

* Mixly		[https://github.com/mu-opensource/MuVisionSensor3-Mixly](https://github.com/mu-opensource/MuVisionSensor3-Mixly)
* Micro:Bit	[https://github.com/mu-opensource/pxt-MuVisionSensor3](https://github.com/mu-opensource/pxt-MuVisionSensor3)

## For more information

Check out the official site [http://www.morpx.com/zn.index.html](http://www.morpx.com/zn.index.html) for links to documentation, issues, and news




release_notes.md
MuVisionSensor V1.2.1
==============================
- 1.bug fixes: now all the examples can works correctly on boards that don't have `SoftwareSerial` library like esp32/8266
- 2.rename from "III" -> "3"
- 3.bug fixes: now MU works correctly after sensor set default
- 4.added new functions: light sensor
- 5.added new enum type for LED: `kLedAll`
- 6.added new library: `MuVisionSensor3_AT.h` for AT commands

MuVisionSensor V1.1.6
==============================
* 1.添加了球算法Label类型：`MU_BALL_TABLE_TENNIS,MU_BALL_TENNIS`
* 2.减少了`begin`函数因通讯异常而导致的程序卡死时间
* 3.弃用了函数`uint8_t begin(void* communication_port, MuVsMode mode)`，改为`uint8_t begin(MuVsUart* communication_port);`和`uint8_t begin(MuVsI2C* communication_port);`
* 4.新增函数`MuVsVisionState* GetVisionState(MuVisionType vision_type)`，获取算法结果缓存地址
* 5.添加了全新的例程、测试程序与相关文档

MuVisionSensor V1.1.5
==============================
* 1.修复了部分Arduino IDE无法找到头文件和examples的问题
* 2.修复了无法从.ZIP加载库文件的问题
* 3.增加了常用函数的注释
* 4.修复了无法设置算法参数的问题
* 5.添加了默认构造函数`MuVisionSensor()`
* 6.将用户常量添加至`MuVisionSensor.h`头文件
* 7.修复了锁定白平衡后一段时间算法结果不稳定的问题

MuVisionSensor V1.1.4 20190117
==============================
* 1.修复了lock位锁定失败而导致的模块死机问题
* 2.修改了LED库函数
* 3.修改了部分枚举变量名




src/mu_vision_sensor_type.h
// Copyright 2018 Morpx Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#ifndef MUVISIONSENSOR_SRC_MUVISIONSENSOR_TYPE_H_
#define MUVISIONSENSOR_SRC_MUVISIONSENSOR_TYPE_H_

#ifdef __cplusplus
 extern "C" {
#endif

#define STREAM_DEBUG_ENABLE 0

typedef unsigned char mu_err_t;

#ifdef BIT
#undef BIT
#endif
#define BIT(x) (0x01<<(x))

#define MU_MAX_RESULT                 1

#define MU_PROTOCOL_VERSION           0x03
#define visionTypeEnumToMacro(v)      (BIT(v-1))

//Error Type
#define MU_OK                         0x00
#define CLIENT_WRITE_TIMEOUT          0x02
#define SERVER_RESPONSE_TIMEOUT       0x03
#define SERVER_RESPONSE_OTHERVISION   0x04
#define MU_ERROR_UNSUPPROT_BAUD       0x10
#define MU_ERROR_UNSUPPROT_PROTOCOL   0x11
#define MU_ERROR_OK                   0xE0
#define MU_ERROR_FAIL                 0xE1
#define MU_ERROR_UNKNOW               0xE2
#define MU_ERROR_TIMEOUT              0xE3
#define MU_ERROR_CHECK_SUM            0xE4
#define MU_ERROR_LENGTH               0xE5
#define MU_ERROR_COMMAND              0xE6
#define MU_ERROR_REG_ADDRESS          0xE7
#define MU_ERROR_REG_VALUE            0xE8
#define MU_ERROR_READ_ONLY            0xE9
//Protocol
#define MU_PROTOCOL_START             0xFF
#define MU_PROTOCOL_END               0xED
#define MU_PROTOCOL_COMMADN_SET       0x01
#define MU_PROTOCOL_COMMADN_GET       0x02
#define MU_PROTOCOL_MESSAGE           0x11

typedef enum {
  kVisionColorDetect  = 1,
  kVisionColorRecog   = 2,
  kVisionBall         = 3,
  kVisionBody         = 5,
  kVisionShapeCard    = 6,
  kVisionTrafficCard  = 7,
  kVisionNumberCard   = 8,
  kVisionMaxType         ,
} MuVsMessageVisionType;
typedef enum {
  kLedClose           = 0,
  kLedRed             = 1,
  kLedGreen           = 2,
  kLedYellow          = 3,
  kLedBlue            = 4,
  kLedPurple          = 5,
  kLedCyan            = 6,
  kLedWhite           = 7,
} MuVsLedColor;
//register address define
typedef enum {
  kRegProtocolVersion = 0x01,
  kRegFirmwareVersion = 0x02,
  kRegRestart         = 0x03,
  kRegSensorConfig1   = 0x04,
  kRegLock            = 0x05,
  kRegLed1            = 0x06,
  kRegLed2            = 0x07,
  kRegLedLevel        = 0x08,
  kRegUart            = 0x09,
  kRegI2C             = 0x0A,
  kRegIO              = 0x0B,
  kRegBle             = 0x0C,
  kRegCameraConfig1   = 0x10,
  kRegFrameCount      = 0x1F,
  kRegVisionId        = 0x20,
  kRegVisionConfig1   = 0x21,
  kRegVisionConfig2   = 0x22,
  kRegParamValue1     = 0x25,
  kRegParamValue2     = 0x26,
  kRegParamValue3     = 0x27,
  kRegParamValue4     = 0x28,
  kRegParamValue5     = 0x29,
  kRegVisionStatus1   = 0x2A,
  kRegVisionStatus2   = 0x2B,
  kRegVisionDetect1   = 0x30,
  kRegVisionDetect2   = 0x31,
  kRegResultNumber    = 0x34,
  kRegResultId        = 0x35,
  kRegReadStatus1     = 0x36,
  kRegResultData1     = 0x40,
  kRegResultData2     = 0x41,
  kRegResultData3     = 0x42,
  kRegResultData4     = 0x43,
  kRegResultData5     = 0x44,
  kRegSn              = 0xD0,
} MuVsRegAddress;

//type define
typedef enum {
  kLed1,
  kLed2,
} MuVsLed;
typedef enum {
  kSerialMode,
  kI2CMode,
} MuVsMode;
typedef enum {
  kBaud9600     = 0x00,
  kBaud19200    = 0x01,
  kBaud38400    = 0x02,
  kBaud57600    = 0x03,
  kBaud115200   = 0x04,
  kBaud230400   = 0x05,
  kBaud460800   = 0x06,
  kBaud921600   = 0x07,
} MuVsBaudrate;
typedef enum {
  kStatus,        // whether the target is detected
  kXValue,        // target horizontal position
  kYValue,        // target vertical position
  kWidthValue,    // target width
  kHeightValue,   // target height
  kLabel,         // target label
  kRValue,        // R channel value
  kGValue,        // G channel value
  kBValue,        // B channel value
} MuVsObjectInf;
typedef enum {
  // for UART mode only
  kCallBackMode = 0,      // u need send a request first, and wait for response
  kDataFlowMode = 1,      // MU will automatically response the result of the vision that u enabled, whether it detected or undetected
  kEventMode    = 2,      // MU can only automatically response the result of the vision that u enabled, which detected target
} MuVsStreamOutputMode;
typedef enum {
  kZoomDefault  = 0,
  kZoom1        = 1,
  kZoom2        = 2,
  kZoom3        = 3,
  kZoom4        = 4,
  kZoom5        = 5,
} MuVsCameraZoom;
typedef enum {
  kFPSNormal        = 0,          // 25FPS mode
  kFPSHigh          = 1,          // 50FPS mode
} MuVsCameraFPS;
typedef enum {
  kAutoWhiteBalance       = 0,    // auto white balance mode
  kLockWhiteBalance       = 1,    // lock white balance with current value, the entire process takes about 100ms
  kWhiteLight             = 2,    // white light mode
  kYellowLight            = 3,    // yellow light mode
} MuVsCameraWhiteBalance;
typedef enum {
  kLevelDefault         = 0,
  kLevelSpeed           = 1,      // speed first mode
  kLevelBalance         = 2,      // balance mode
  kLevelAccuracy        = 3,      // accuracy first mode
} MuVsVisionLevel;

// register type
typedef union {
  struct {
    unsigned char reserve0 : 2;
    unsigned char default_setting :1;  // set 1 reset all config
  };
  unsigned char sensor_config_reg_value;
} MuVsSensorConfig1;
typedef union {
  struct {
    MuVsBaudrate baudrate :3;
  };
  unsigned char uart_reg_value;
} MuVsUartConfig;
typedef union {
  struct {
    unsigned char manual:1;
    MuVsLedColor detected_color:3;
    unsigned char hold:1;
    MuVsLedColor undetected_color:3;
  };
  unsigned char led_reg_value;
} MuVsLedConfig;
typedef union {
  struct {
    MuVsCameraZoom zoom:3;
    unsigned char rotate:1;
    MuVsCameraFPS fps:1;
    MuVsCameraWhiteBalance white_balance:2;
    unsigned char calibration:1;
  };
  unsigned char camera_reg_value;
} MuVsCameraConfig1;
typedef union {
  struct {
    unsigned char status :1;
    unsigned char default_setting :1;  // set 1 to reset vision configuration
    MuVsStreamOutputMode output_mode :2;
    MuVsVisionLevel level :2;
    unsigned char reserve6 :1;
    unsigned char output_enable :1;
  };
  unsigned char vision_config_reg_value;
} MuVsVisionConfig1;
typedef struct {
  unsigned char frame;
  unsigned char detect;
  struct {
    union {
      unsigned char result_data1;
      unsigned char x_value;
      unsigned char color_r_value;
      unsigned char top_x_value;
    };
    union {
      unsigned char result_data2;
      unsigned char y_value;
      unsigned char color_g_value;
      unsigned char top_y_value;
    };
    union {
      unsigned char result_data3;
      unsigned char width;
      unsigned char color_b_value;
      unsigned char bot_x_value;
    };
    union {
      unsigned char result_data4;
      unsigned char height;
      unsigned char bot_y_value;
    };
    union {
      unsigned char result_data5;
      unsigned char color;
      unsigned char lable;
    };
  } vision_result[MU_MAX_RESULT];
} MuVsVisionState;

#ifdef __cplusplus
 }
#endif

#endif /* ARDUINO_LIB_MUVISIONSENSOR_SRC_MUVISIONSENSOR_TYPE_H_ */


src/mu_vision_sensor_protocol.h
// Copyright 2018 Morpx Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#ifndef ARDUINO_LIB_MUVISIONSENSOR_SRC_MU_VISION_SENSOR_PROTOCOL_H_
#define ARDUINO_LIB_MUVISIONSENSOR_SRC_MU_VISION_SENSOR_PROTOCOL_H_

#include <stdlib.h>
#include <stdint.h>
#include "mu_vision_sensor_type.h"

#ifdef __cplusplus
extern "C" {
#endif

/*
 * MU vision sensor hardware interface function.
 * Get serial data number.
 * Return: byte number in serial buffer.
 */
typedef uint32_t (*MuVsUartAvailable)(void);
/*
 * MU vision sensor hardware interface function.
 * Read serial data.
 * Input :temp: unsigned char buffer pointer
 *        length: max buffer length
 * Return: actual buffer length,if return number less then (length),it means read time out.
 */
typedef uint32_t (*MuVsUartRead)(uint8_t* temp, uint8_t length);

/*
 * MU vision sensor hardware interface function.
 * Write serial data.
 * Input :temp: write unsigned char buffer pointer
 *        length: max buffer length
 * Return: actual buffer length,if return number less then (length),it means serial port write buffer is full.
 */
typedef uint32_t (*MuVsUartWrite)(uint8_t* temp, uint8_t length);

typedef struct {
  MuVsUartAvailable uart_avalible;
  MuVsUartRead uart_read;
  MuVsUartWrite uart_write;
} Uart;

mu_err_t mu_vs_uart_init(MuVsUartAvailable func_uart_available,
                         MuVsUartRead func_uart_read,
                         MuVsUartWrite func_uart_write);
mu_err_t mu_vs_uart_set(const uint8_t mu_address, const uint8_t reg_address, const uint8_t value);
mu_err_t mu_vs_uart_get(const uint8_t mu_address, const uint8_t reg_address, uint8_t* value);
mu_err_t mu_vs_uart_read(uint8_t* mu_address,
                         MuVsMessageVisionType* vision_type,
                         MuVsVisionState* vision_state);

#ifdef __cplusplus
}
#endif

#endif /* ARDUINO_LIB_MUVISIONSENSOR_SRC_MU_VISION_SENSOR_PROTOCOL_H_ */


src/MuVisionSensor.h
// Copyright 2018 Morpx Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#ifndef MUVISIONSENSOR_SRC_MUVISIONSENSOR_H_
#define MUVISIONSENSOR_SRC_MUVISIONSENSOR_H_

#include "mu_vision_sensor_interface.h"
#include "mu_vision_sensor_i2c_hw_interface.h"
#include "mu_vision_sensor_uart_interface.h"

/*
 * MuVisionType
 * Bit: |          3         |          2         |            1             |           0         |
 *      | VISION_LINE_DETECT | VISION_BALL_DETECT | VISION_COLOR_RECOGNITION | VISION_COLOR_DETECT |
 *      |           7            |             6              |             5            |          4         |
 *      | VISION_NUM_CARD_DETECT | VISION_TRAFFIC_CARD_DETECT | VISION_SHAPE_CARD_DETECT | VISION_BODY_DETECT |
 */
typedef unsigned short MuVisionType;
typedef unsigned char MuLightSensorType;

// MuVisionType: Vision Type User Input
#define VISION_COLOR_DETECT           (MuVisionType)visionTypeEnumToMacro(kVisionColorDetect)
#define VISION_COLOR_RECOGNITION      (MuVisionType)visionTypeEnumToMacro(kVisionColorRecog)
#define VISION_BALL_DETECT            (MuVisionType)visionTypeEnumToMacro(kVisionBall)
#define VISION_BODY_DETECT            (MuVisionType)visionTypeEnumToMacro(kVisionBody)
#define VISION_SHAPE_CARD_DETECT      (MuVisionType)visionTypeEnumToMacro(kVisionShapeCard)
#define VISION_TRAFFIC_CARD_DETECT    (MuVisionType)visionTypeEnumToMacro(kVisionTrafficCard)
#define VISION_NUM_CARD_DETECT        (MuVisionType)visionTypeEnumToMacro(kVisionNumberCard)
#define VISION_ALL                    (MuVisionType)(visionTypeEnumToMacro(kVisionMaxType)-1)
// MuLightSensorType: light sensor user input
#define LS_PROXIMITY_ENABLE           BIT(0)
#define LS_AMBIENT_LIGHT_ENABLE       BIT(1)
#define LS_COLOR_ENABLE               BIT(2)
#define LS_GESTURE_ENABLE             BIT(3)

// Card Type
// Vision Shape Card
#define MU_SHAPE_CARD_TICK            0x01U
#define MU_SHAPE_CARD_CROSS           0x02U
#define MU_SHAPE_CARD_CIRCLE          0x03U
#define MU_SHAPE_CARD_SQUARE          0x04U
#define MU_SHAPE_CARD_TRIANGLE        0x05U
// Vision Traffic Card
#define MU_TRAFFIC_CARD_FORWARD       0x01U
#define MU_TRAFFIC_CARD_LEFT          0x02U
#define MU_TRAFFIC_CARD_RIGHT         0x03U
#define MU_TRAFFIC_CARD_TURN_AROUND   0x04U
#define MU_TRAFFIC_CARD_PARK          0x05U
// Vision Color Type
#define MU_COLOR_UNKNOWN              0x00U
#define MU_COLOR_BLACK                0x01U
#define MU_COLOR_WHITE                0x02U
#define MU_COLOR_RED                  0x03U
#define MU_COLOR_YELLOW               0x04U
#define MU_COLOR_GREEN                0x05U
#define MU_COLOR_CYAN                 0x06U
#define MU_COLOR_BLUE                 0x07U
#define MU_COLOR_PURPLE               0x08U
// Vision Ball Type
#define MU_BALL_TABLE_TENNIS          0x01U
#define MU_BALL_TENNIS                0x02U


class MuVisionSensor {
 public:
  /**
    * @brief  construct class MU vision sensor.
    * @param  address: MU vision sensor device address.
    * @retval none
    */
  MuVisionSensor(uint32_t address = 0x60);
  virtual ~MuVisionSensor();

  /**
    * @ TODO WARNING this function may delete in later version, please use `begin(communication_port)` instead.
    * @brief  MU vision sensor begin.
    * @param  communication_port: MuVsI2C(i2c) or MuVsUart(uart).
    * @param  mode: kSerialMode
    *               kI2CMode
    * @retval MU_OK: begin success.
    *         other: protocol assert fail.
    */
  uint8_t __attribute__ ((deprecated("\n***WARNING*** function `begin(communication_port, mode)` has been deprecated, and may delete in later version, please use `begin(communication_port)` instead.")))
          begin(void* communication_port,
                MuVsMode mode);
  /**
    * @brief  MU vision sensor begin.
    * @param  communication_port: MuVsUart(uart).
    * @retval MU_OK: begin success.
    *         other: protocol assert fail.
    */
  uint8_t begin(MuUart::hw_port_t communication_port);
  /**
    * @brief  MU vision sensor begin.
    * @param  communication_port: MuVsI2C(i2c).
    * @retval MU_OK: begin success.
    *         other: protocol assert fail.
    */
  uint8_t begin(MuVsI2C* communication_port);

  // Based interface
  /**
    * @brief  begin vision.
    * @param  vision_type: vision type.
    * @retval MU_OK:  success
    *         other:  error
    */
  uint8_t VisionBegin(MuVisionType);
  /**
    * @brief  end vision.
    * @param  vision_type: vision type.
    * @retval MU_OK:  success
    *         other:  error
    */
  uint8_t VisionEnd(MuVisionType);
  /**
    * @brief  get vision result data, this function will update vision
    *         result automatically.
    * @param  vision_type: vision type.
    * @param  object_inf:  object information
    * @retval information value
    */
  int GetValue(MuVisionType vision_type,
               MuVsObjectInf object_inf);
  /**
    * @brief  write vision parameter.
    * @param  vision_type: vision type.
    * @param  object_inf:  object information
    * @param  value:  value
    * @retval MU_OK:  success
    *         other:  error
    */
  int SetValue(MuVisionType vision_type,
               MuVsObjectInf object_inf,
               uint8_t value) {
    return write(vision_type, object_inf, value);
  }
  /**
    * @brief  get vision result buffer pointer.
    *         this function WILL NOT update vision result, please use
    *         function `UpdateResult([vision_type])` or
    *         `GetValue([vision_type], kStatus)` to update vision result before this function
    * @param  vision_type: vision type.
    * @retval vision result buffer pointer,
    *         return `nullptr` if the vision type is not `begin` or not supported
    */
  MuVsVisionState* GetVisionState(MuVisionType vision_type);

  // Advance interface
  /**
    * @brief  update result data from vision sensor, must used after
    *         VisionBegin or VisionSetStatus.
    * @param  vision_type vision type.
    * @param  wait_all_result  true: return if get all input vision type
    *                           false: return if get one of the input vision type
    * @retval the vision type which have been updated
    */
  MuVisionType UpdateResult(MuVisionType vision_type,
                            bool wait_all_result = true);
  /**
    * @brief  write vision parameter.
    * @param  vision_type vision type.
    * @param  object_inf  object information
    * @param  value  value
    * @retval MU_OK  success
    *         other  error
    */
  uint8_t write(MuVisionType, MuVsObjectInf, uint8_t value);
  /**
    * @brief  read result data.
    * @param  vision_type vision type.
    * @param  object_inf  object information
    * @retval information value
    */
  uint8_t read(MuVisionType vision_type,
               MuVsObjectInf object_inf,
               uint8_t result_num = 1);

  // Sensor functions
  //!< @brief  restart MU vision sensor
  uint8_t SensorSetRestart(void);
  //!< @brief  set all register to default value(include baud rate)
  uint8_t SensorSetDefault(void);

  // LED functions
  /**
    * @brief  set led.
    * @param  led led type.
    * @param  manual vision type.
    * @param  hold  object information
    * @retval MU_OK  success
    *         other  error
    */
  uint8_t LedSetMode(MuVsLed led, bool manual, bool hold);
  /**
    * @brief  set led color.
    * @param  led led type.
    * @param  detected_color led color while sensor detected target.
    * @param  undetected_color led color while sensor undetected target.
    * @param  level  led brightness, form 0(close) to 15
    * @retval MU_OK  success
    *         other  error
    */
  uint8_t LedSetColor(MuVsLed led,
                      MuVsLedColor detected_color,
                      MuVsLedColor undetected_color,
                      uint8_t level = 1);

  // Camera functions
  /**
    * @brief  set camera zoom.
    * @param  camera zoom value.
    * @retval MU_OK  success
    *         other  error
    */
  uint8_t CameraSetZoom(MuVsCameraZoom);
  /**
    * @brief  rotate camera.
    * @param  enable true: rotate camera.
    *                 false: default
    * @retval MU_OK  success
    *         other  error
    */
  uint8_t CameraSetRotate(bool enable);
  /**
    * @brief  set camera zoom.
    * @param  camera FPS type.
    * @retval MU_OK:  success
    *         other:  error
    */
  uint8_t CameraSetFPS(MuVsCameraFPS);
  /**
    * @brief  set camera white balance.
    * @param  camera white balance type.
    * @retval MU_OK  success
    *         other  error
    */
  uint8_t CameraSetAwb(MuVsCameraWhiteBalance);
  /**
    * @brief  get camera zoom value.
    * @retval camera zoom value
    */
  MuVsCameraZoom CameraGetZoom(void);
  /**
    * @brief  get camera AWB type.
    * @retval camera AWB type
    */
  MuVsCameraWhiteBalance CameraGetAwb(void);
  /**
    * @brief  get camera rotate state.
    * @retval camera rotate state
    */
  bool CameraGetRotate(void);
  /**
    * @brief  get camera FPS type.
    * @retval camera FPS type
    */
  MuVsCameraFPS CameraGetFPS(void);

  // Uart functions
  /**
    * @brief  set UART baud rate.
    * @param  UART baud rate.
    * @retval MU_OK:  success
    *         other:  error
    */
  uint8_t UartSetBaudrate(MuVsBaudrate);

  // Vision functions
  /**
    * @brief  set vision status.
    * @param  vision_type: vision type
    * @param  enable: true: enable vision
    *                 false: disable vision
    * @retval MU_OK:  success
    *         other:  error
    */
  uint8_t VisionSetStatus(MuVisionType vision_type, bool enable);
  /**
    * @brief  set vision status.
    * @param  vision_type: vision type
    * @param  mode: output mode
    * @retval MU_OK:  success
    *         other:  error
    */
  uint8_t VisionSetOutputMode(MuVsStreamOutputMode mode);
  /**
    * @brief  output enable.
    * @param  vision_type: vision type
    * @param  status: true: start output
    *                 false: stop output
    * @retval MU_OK:  success
    *         other:  error
    */
  uint8_t VisionSetOutputEnable(MuVisionType vision_type, bool status);
  /**
    * @brief  set vision configure to default value.
    * @param  vision_type: vision type
    * @retval MU_OK:  success
    *         other:  error
    */
  uint8_t VisionSetDefault(MuVisionType vision_type);
  /**
    * @brief  set vision level.
    * @param  vision_type: vision type
    * @param  level: vision level
    * @retval MU_OK:  success
    *         other:  error
    */
  uint8_t VisionSetLevel(MuVisionType vision_type,
                         MuVsVisionLevel level);
  /**
    * @brief  get vision status.
    * @retval vision status
    */
  bool VisionGetStatus(MuVisionType vision_type);
  /**
    * @brief  get vision level.
    * @retval vision level
    */
  MuVsVisionLevel VisionGetLevel(MuVisionType vision_type);
  /**
    * @brief  get vision output mode.
    * @retval vision output mode
    */
  MuVsStreamOutputMode VisionGetOutputMode(void);

  /*
   * @brief Light sensor function begin.
   * @param[in] ls_type Function type
   * @retval MU_OK: success
   *         otherwise: failure.
   */
  uint8_t LsBegin(MuLightSensorType ls_type);
  /*
   * @brief Light sensor function end.
   * @param[in] ls_type Function type
   * @retval MU_OK: success
   *         otherwise: failure.
   */
  uint8_t LsEnd(MuLightSensorType ls_type);
  /*
   * @brief Light snesor set sensitivity.
   * @retval MU_OK: success
   *         otherwise: failure.
   */
  uint8_t LsSetSensitivity(MuVsLsSensitivity);
  /*
   * @brief Enable white balance correction.
   * @retval MU_OK: success
   *         otherwise: failure.
   */
  uint8_t LsWhiteBalanceEnable();
  /*
   * @brief Read proximity data.
   * @retval proximity data, 0~255.
   */
  uint8_t LsReadProximity();
  /*
   * @brief Read ambient light sensor data.
   * @retval ambient light sensor data, 0~65536.
   */
  uint16_t LsReadAmbientLight();
  /*
   * @brief Read light sensor color data.
   * @param[in] color_t kLsColorLabel:      get color label
   *                    kLsColorRed:        get RGB R value(0~255)
   *                    kLsColorGreen:      get RGB G value(0~255)
   *                    kLsColorBlue:       get RGB B value(0~255)
   *                    kLsColorHue:        get HSV H value(0~360)
   *                    kLsColorSaturation: get HSV S value(0~255)
   *                    kLsColorValue:      get HSV V value(0~255)
   * @retval Light sensor detected color data value
   */
  uint16_t LsReadColor(MuVsLsColorType color_t);
  /*
   * @brief Read light sensor color raw data.
   * @param[in] color_t kLsRawColorRed:   get raw color R value
   *                    kLsRawColorGreen: get raw color G value
   *                    kLsRawColorBlue:  get raw color B value
   * @retval Light sensor color raw data value
   */
  uint16_t LsReadRawColor(MuVsLsRawColorType color_t);
  /*
   * @brief Read gesture sensor data.
   * @retval Gesture witch MU detected.
   */
  MuVsLsGesture LsReadGesture();


  MuVisionSensor(const MuVisionSensor&) = delete;
  MuVisionSensor& operator=(const MuVisionSensor &) = delete;

 private:
  uint8_t SensorLockReg(bool lock);
  uint8_t ProtocolVersionCheck();
  MuVisionType UartUpdateResult(MuVisionType vision_type, bool wait_all_result);
  bool malloc_vision_buffer(MuVsMessageVisionType);
  bool free_vision_buffer(MuVsMessageVisionType);

  uint8_t address_ = 0;
  MuVsMode mode_ = (MuVsMode)0xFF;
  MuVsMethod* mu_vs_method = nullptr;
  MuVsStreamOutputMode output_mode_ = kCallBackMode;
  MuVsVisionState *vision_state_[kVisionMaxType-1] = {nullptr};
};


#endif /* ARDUINO_LIB_MUVISIONSENSOR_SRC_MUVISIONSENSOR_H_ */
